src/chat.rs

```diff
@@ -152,6 +180,130 @@ impl ChatId {
+                    }
+                }
+                Chattype::Undefined => bail!("set protection: undefined group type"),
+            },
+            ProtectionStatus::Unprotected => {}
+        };
+
+        context
+            .sql
+            .execute(
+                "UPDATE chats SET protected=? WHERE id=?;",
+                paramsv![protect, self],
+            )
+            .await?;
+
+        context.emit_event(EventType::ChatModified(self));
+
+        // make sure, the receivers will get all keys
+        reset_gossiped_timestamp(context, self).await?;
+
+        Ok(())
+    }
+
+    /// Returns a stock message saying that protection status has changed.
+    pub(crate) async fn get_protection_msg(
+        self,
+        context: &Context,
+        protect: ProtectionStatus,
+        from_id: u32,
+    ) -> String {
+        context
+            .stock_system_msg(
+                match protect {
+                    ProtectionStatus::Protected => StockMessage::ProtectionEnabled,
+                    ProtectionStatus::Unprotected => StockMessage::ProtectionDisabled,
+                },
+                "",
+                "",
+                from_id,
+            )
+            .await
+    }
+
+    /// Send protected status message to the chat.
+    ///
+    /// This sends the message with the protected status change to the chat,
+    /// notifying the user on this device as well as the other users in the chat.
+    /// If `promoted` is false this means the chat only exists on this device so far
+    /// and does not need to be sent out.
+    /// In this case an local info message is added to the chat.
+    pub(crate) async fn add_protection_msg(
+        self,
+        context: &Context,
+        protect: ProtectionStatus,
+        promoted: bool,
+        from_id: u32,
+    ) -> Result<(), Error> {
+        let msg_text = self.get_protection_msg(context, protect, from_id).await;
+
+        if promoted {
+            let mut msg = Message::default();
+            msg.viewtype = Viewtype::Text;
+            msg.text = Some(msg_text);
+            msg.param.set_cmd(match protect {
+                ProtectionStatus::Protected => SystemMessage::ChatProtectionEnabled,
+                ProtectionStatus::Unprotected => SystemMessage::ChatProtectionDisabled,
+            });
+            send_msg(context, self, &mut msg).await?;
+        } else {
+            add_info_msg(context, self, msg_text).await;
+        }
+
+        Ok(())
+    }
+
+    /// Sets protection and sends or adds a message.
+    pub async fn set_protection(
+        self,
+        context: &Context,
+        protect: ProtectionStatus,
+    ) -> Result<(), Error> {
+        ensure!(!self.is_special(), "set protection: invalid chat-id.");
+
+        let chat = Chat::load_from_db(context, self).await?;
+
+        if let Err(e) = self.inner_set_protection(context, protect).await {
+            error!(context, "{}", e); // make error user-visible
```

Comment:
Didn't we decide to stop using these?  Should we not send an device message to the chat saying enable protection failed for whatever reason?

Reply:
currently, errors still go through error!() and display as a toast or so.

ux-wise, i also do not think, that these sort of errors should go to a device-chat, it is an error after a concrete user-action, so we could just display an error-alert, however, we are not there and that is true for some other places as well. getting rid of error! is out of scope of this pr :)